File: ios/Calc/Models/RomStorage.swift
//
//  RomStorage.swift
//  Calc
//
//  Handles ROM file storage and persistence.
//  Saves ROMs to documents directory and remembers the last loaded ROM.
//

import Foundation
import os.log

/// Handles ROM file storage and persistence
struct RomStorage {
    private static let logger = Logger(subsystem: "com.calc.emulator", category: "RomStorage")
    private static let romFilenameKey = "lastRomFilename"
    private static let romDirName = "roms"

    /// Get the ROM storage directory, creating it if needed
    private static var romDirectory: URL? {
        let fileManager = FileManager.default
        guard let documentsDir = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first else {
            logger.error("Could not find documents directory")
            return nil
        }

        let romDir = documentsDir.appendingPathComponent(romDirName, isDirectory: true)

        if !fileManager.fileExists(atPath: romDir.path) {
            do {
                try fileManager.createDirectory(at: romDir, withIntermediateDirectories: true)
            } catch {
                logger.error("Failed to create ROM directory: \(error.localizedDescription)")
                return nil
            }
        }

        return romDir
    }

    /// Save ROM data to storage and remember it
    /// - Returns: true if saved successfully
    static func saveRom(_ data: Data, originalName: String) -> Bool {
        guard let romDir = romDirectory else { return false }

        let safeFilename = sanitizeFilename(originalName)
        let romURL = romDir.appendingPathComponent(safeFilename)

        do {
            try data.write(to: romURL)
            UserDefaults.standard.set(safeFilename, forKey: romFilenameKey)
            logger.info("Saved ROM: \(safeFilename) (\(data.count) bytes)")
            return true
        } catch {
            logger.error("Failed to save ROM: \(error.localizedDescription)")
            return false
        }
    }

    /// Load the previously saved ROM if available
    /// - Returns: ROM data and filename, or nil if no saved ROM
    static func loadSavedRom() -> (data: Data, filename: String)? {
        guard let filename = UserDefaults.standard.string(forKey: romFilenameKey),
              let romDir = romDirectory else {
            return nil
        }

        let romURL = romDir.appendingPathComponent(filename)

        guard FileManager.default.fileExists(atPath: romURL.path) else {
            logger.warning("Saved ROM file not found: \(filename)")
            clearSavedRom()
            return nil
        }

        do {
            let data = try Data(contentsOf: romURL)
            logger.info("Loaded saved ROM: \(filename) (\(data.count) bytes)")
            return (data, filename)
        } catch {
            logger.error("Failed to load saved ROM: \(error.localizedDescription)")
            return nil
        }
    }

    /// Check if there's a saved ROM available
    static func hasSavedRom() -> Bool {
        guard let filename = UserDefaults.standard.string(forKey: romFilenameKey),
              let romDir = romDirectory else {
            return false
        }

        let romURL = romDir.appendingPathComponent(filename)
        return FileManager.default.fileExists(atPath: romURL.path)
    }

    /// Clear the saved ROM preference (does not delete the file)
    static func clearSavedRom() {
        UserDefaults.standard.removeObject(forKey: romFilenameKey)
    }

    // MARK: - Emulator State Persistence

    private static let stateFilename = "emulator.state"

    /// Get the state file URL
    private static var stateFileURL: URL? {
        let fileManager = FileManager.default
        guard let documentsDir = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first else {
            return nil
        }
        return documentsDir.appendingPathComponent(stateFilename)
    }

    /// Save emulator state to storage
    /// - Returns: true if saved successfully
    static func saveState(_ data: Data) -> Bool {
        guard let stateURL = stateFileURL else { return false }

        do {
            try data.write(to: stateURL)
            logger.info("Saved emulator state: \(data.count) bytes")
            return true
        } catch {
            logger.error("Failed to save emulator state: \(error.localizedDescription)")
            return false
        }
    }

    /// Load emulator state from storage
    /// - Returns: State data, or nil if no saved state
    static func loadState() -> Data? {
        guard let stateURL = stateFileURL,
              FileManager.default.fileExists(atPath: stateURL.path) else {
            logger.debug("No saved emulator state found")
            return nil
        }

        do {
            let data = try Data(contentsOf: stateURL)
            logger.info("Loaded emulator state: \(data.count) bytes")
            return data
        } catch {
            logger.error("Failed to load emulator state: \(error.localizedDescription)")
            return nil
        }
    }

    /// Check if there's a saved emulator state
    static func hasSavedState() -> Bool {
        guard let stateURL = stateFileURL else { return false }
        return FileManager.default.fileExists(atPath: stateURL.path)
    }

    /// Delete the saved emulator state
    static func clearSavedState() {
        guard let stateURL = stateFileURL,
              FileManager.default.fileExists(atPath: stateURL.path) else {
            return
        }

        do {
            try FileManager.default.removeItem(at: stateURL)
            logger.info("Cleared saved emulator state")
        } catch {
            logger.error("Failed to clear emulator state: \(error.localizedDescription)")
        }
    }

    /// Sanitize filename for safe storage
    private static func sanitizeFilename(_ name: String) -> String {
        // Remove path separators and problematic characters
        let sanitized = name.replacingOccurrences(of: "[/\\\\:*?\"<>|]", with: "_", options: .regularExpression)
        // Ensure it has .rom extension if no extension
        return sanitized.contains(".") ? sanitized : "\(sanitized).rom"
    }
}

File: android/app/src/main/java/com/calc/emulator/RomStorage.kt
package com.calc.emulator

import android.content.Context
import android.util.Log
import java.io.File

/**
 * Handles ROM file storage and persistence.
 * Saves ROMs to internal storage and remembers the last loaded ROM.
 */
object RomStorage {
    private const val TAG = "RomStorage"
    private const val PREFS_NAME = "emulator_prefs"
    private const val KEY_ROM_FILENAME = "last_rom_filename"
    private const val ROM_DIR = "roms"
    private const val STATE_FILENAME = "emulator.state"

    /**
     * Save ROM data to internal storage and remember it.
     * @return true if saved successfully
     */
    fun saveRom(context: Context, romBytes: ByteArray, originalName: String): Boolean {
        return try {
            val romDir = File(context.filesDir, ROM_DIR)
            if (!romDir.exists()) {
                romDir.mkdirs()
            }

            // Use a sanitized filename
            val safeFilename = sanitizeFilename(originalName)
            val romFile = File(romDir, safeFilename)
            romFile.writeBytes(romBytes)

            // Remember the filename
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                .edit()
                .putString(KEY_ROM_FILENAME, safeFilename)
                .apply()

            Log.i(TAG, "Saved ROM: $safeFilename (${romBytes.size} bytes)")
            true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save ROM", e)
            false
        }
    }

    /**
     * Load the previously saved ROM if available.
     * @return ROM bytes and filename, or null if no saved ROM
     */
    fun loadSavedRom(context: Context): Pair<ByteArray, String>? {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val filename = prefs.getString(KEY_ROM_FILENAME, null) ?: return null

        val romFile = File(File(context.filesDir, ROM_DIR), filename)
        if (!romFile.exists()) {
            Log.w(TAG, "Saved ROM file not found: $filename")
            clearSavedRom(context)
            return null
        }

        return try {
            val bytes = romFile.readBytes()
            Log.i(TAG, "Loaded saved ROM: $filename (${bytes.size} bytes)")
            Pair(bytes, filename)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load saved ROM", e)
            null
        }
    }

    /**
     * Check if there's a saved ROM available.
     */
    fun hasSavedRom(context: Context): Boolean {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val filename = prefs.getString(KEY_ROM_FILENAME, null) ?: return false
        val romFile = File(File(context.filesDir, ROM_DIR), filename)
        return romFile.exists()
    }

    /**
     * Clear the saved ROM preference (does not delete the file).
     */
    fun clearSavedRom(context: Context) {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .edit()
            .remove(KEY_ROM_FILENAME)
            .apply()
    }

    /**
     * Save emulator state to internal storage.
     * @return true if saved successfully
     */
    fun saveState(context: Context, stateBytes: ByteArray): Boolean {
        return try {
            val stateFile = File(context.filesDir, STATE_FILENAME)
            stateFile.writeBytes(stateBytes)
            Log.i(TAG, "Saved emulator state: ${stateBytes.size} bytes")
            true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save emulator state", e)
            false
        }
    }

    /**
     * Load emulator state from internal storage.
     * @return State bytes, or null if no saved state
     */
    fun loadState(context: Context): ByteArray? {
        val stateFile = File(context.filesDir, STATE_FILENAME)
        if (!stateFile.exists()) {
            Log.d(TAG, "No saved emulator state found")
            return null
        }

        return try {
            val bytes = stateFile.readBytes()
            Log.i(TAG, "Loaded emulator state: ${bytes.size} bytes")
            bytes
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load emulator state", e)
            null
        }
    }

    /**
     * Check if there's a saved emulator state.
     */
    fun hasSavedState(context: Context): Boolean {
        return File(context.filesDir, STATE_FILENAME).exists()
    }

    /**
     * Delete the saved emulator state.
     */
    fun clearSavedState(context: Context) {
        val stateFile = File(context.filesDir, STATE_FILENAME)
        if (stateFile.exists()) {
            stateFile.delete()
            Log.i(TAG, "Cleared saved emulator state")
        }
    }

    private fun sanitizeFilename(name: String): String {
        // Remove path separators and other potentially problematic characters
        val sanitized = name.replace(Regex("[/\\\\:*?\"<>|]"), "_")
        // Ensure it has .rom extension if it doesn't have one
        return if (sanitized.contains('.')) sanitized else "$sanitized.rom"
    }
}
